File: ga/4omini_result_2024_7/result/1968G1_mini/27_1968G1.c
use rand::Rng;
use std::io::{self, BufRead};
use std::time::SystemTime;

const N: usize = 200005;
const MOD: usize = 1_000_000_007;

fn random_base() -> usize {
    let mut rng = rand::thread_rng();
    rng.gen_range(200..MOD) // Generates a random base in the range [200, MOD-1]
}

fn power(a: usize, mut b: usize) -> usize {
    let mut r = 1;
    let mut a = a;

    while b > 0 {
        if b % 2 == 1 {
            r = r * a % MOD;
        }
        a = a * a % MOD;
        b /= 2;
    }
    r
}

fn check(k: usize, n: usize, hash: &[usize], base_pow: &[usize], s: &[usize]) -> usize {
    let mut vo = 0;
    let mut i = 1;
    
    while i + k - 1 <= n {
        let hash_sub = (hash[i + k - 1] + MOD - hash[i - 1]) % MOD;
        if hash_sub == (hash[k] * base_pow[i - 1]) % MOD {
            vo += 1;
            i += k;
        } else {
            i += 1;
        }
    }
    vo
}

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let mut lines = reader.lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    let mut rng = rand::thread_rng();
    
    for _ in 0..t {
        let params: Vec<usize> = lines.next().unwrap().unwrap().split_whitespace()
            .map(|x| x.parse().unwrap()).collect();
        let (n, l, r) = (params[0], params[1], params[2]);
        
        let s: Vec<char> = lines.next().unwrap().unwrap().chars().collect();
        
        let mut base_pow = vec![1; n + 1];
        let mut hash = vec![0; n + 1];
        
        let base = random_base();
        
        for i in 1..=n {
            base_pow[i] = base_pow[i - 1] * base % MOD;
            hash[i] = (hash[i - 1] * base + (s[i - 1] as usize)) % MOD; // s is 0-indexed
        }

        let mut lower = 0;
        let mut upper = n + 1;

        while upper - lower > 1 {
            let mid = lower + (upper - lower) / 2;
            if check(mid, n, &hash, &base_pow, &s) >= l {
                lower = mid;
            } else {
                upper = mid;
            }
        }
        
        print!("{}", lower);

        let mut yy = check(lower, n, &hash, &base_pow, &s);
        
        for i in l + 1..=r {
            if i > yy {
                upper = lower + 1;
                lower = 0;

                while upper - lower > 1 {
                    let mid = lower + (upper - lower) / 2;
                    if check(mid, n, &hash, &base_pow, &s) >= i {
                        lower = mid;
                    } else {
                        upper = mid;
                    }
                }
            }
            print!(" {}", lower);
        }
        println!();
    }
}

File: ga/glm4_result_2024_6/result/1968G1_glm/27_1968G1.rs
use rand::Rng;
use std::time::SystemTime;

const N: usize = 200005;
const MOD: i64 = 1000000007;

fn rand1() -> u32 {
    rand::random::<u32>() as u32
}

fn rand2(l: u32, r: u32) -> u32 {
    rand::random::<u32>() % (r - l + 1) + l
}

fn power(a: i64, b: i64) -> i64 {
    if b == 0 {
        return 1;
    }
    let r = power(a, b / 2);
    if b % 2 == 1 {
        return r * r % MOD * a % MOD;
    }
    r * r % MOD
}

fn check(k: usize, hash: &[i64], base_pow: &[i64]) -> usize {
    let mut vo = 0;
    let mut i = 1;
    while i + k - 1 <= hash.len() {
        if (hash[i + k - 1] + MOD - hash[i - 1]) % MOD == hash[k] * base_pow[i - 1] % MOD {
            vo += 1;
            i += k;
        } else {
            i += 1;
        }
    }
    vo
}

fn main() {
    let mut rng = rand::thread_rng();
    let base: u32 = rng.gen_range(200..MOD as u32);
    let mut base_pow = vec![0; N + 1];
    base_pow[0] = 1;
    for i in 1..=N {
        base_pow[i] = base_pow[i - 1] * base as i64 % MOD;
    }

    let t = rand::random::<usize>() % 100 + 1;
    for _ in 0..t {
        let n = rand::random::<usize>() % 100 + 1;
        let l = rand::random::<usize>() % n + 1;
        let r = rand::random::<usize>() % n + 1;
        let s = (0..n).map(|_| rand::random::<char>()).collect::<String>();

        let mut hash = vec![0; n + 1];
        hash[0] = 0;
        for i in 1..=n {
            hash[i] = hash[i - 1] * base as i64 % MOD + s.chars().nth(i - 1).unwrap() as i64 % MOD;
            hash[i] %= MOD;
        }

        let mut lower = 0;
        let mut upper = n + 1;
        while upper - lower > 1 {
            let mid = lower + (upper - lower) / 2;
            if check(mid, &hash, &base_pow) >= l {
                lower = mid;
            } else {
                upper = mid;
            }
        }
        println!("{}", lower);

        for i in l + 1..=r {
            if i > check(lower, &hash, &base_pow) {
                upper = lower + 1;
                lower = 0;
                while upper - lower > 1 {
                    let mid = lower + (upper - lower) / 2;
                    if check(mid, &hash, &base_pow) >= i {
                        lower = mid;
                    } else {
                        upper = mid;
                    }
                }
            }
            println!("{}", lower);
        }
    }
}